Feature Flag evaluation event (#1440)
open-telemetry/semantic-conventions: docs/attributes-registry/feature-flag.md

| `feature_flag.key`            | string | The unique identifier of the feature flag.                                                                              | `logo-color`                           | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `feature_flag.provider.id`    | string | The name of the service provider that performs the flag evaluation.                                                     | `Flag Manager`                         | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `feature_flag.value`          | string | The evaluated value of the feature flag.                                                                                | `#ff0000`; `1`; `true`                 | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `feature_flag.variant`        | string | SHOULD be a semantic identifier for a value. If one is unavailable, a stringified version of the value can be used. [1] | `red`; `true`; `on`                    | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

> | `feature_flag.version`        | string | The version of the ruleset used during the evaluation. Could use a hash if a version is not available.                  | `1`; `01ABCDEF`                        | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer: > Could use a hash if a version is not available.

Why hash specifically? Can't it be any stable value representing the state of the ruleset?

| `error` | The resolved value was the result of an error. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `split` | The resolved value was the result of pseudorandom assignment. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `stale` | The resolved value is non-authoritative or possibly out of date | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `static` | The resolved value is static (no dynamic evaluation). | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `targeting_match` | The resolved value was the result of a dynamic evaluation, such as a rule or specific user-targeting. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `unknown` | The reason for the resolved value could not be determined. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

## Deprecated Feature Flag Attributes

Describes deprecated Feature Flag attributes.

| Attribute | Type | Description | Examples | Stability |
|---|---|---|---|---|

> | <a id="feature-flag-provider-name" href="#feature-flag-provider-name">`feature_flag.provider_name`</a> | string | Deprecated, use `feature_flag.provider.id` instead. | `Flag Manager` | ![Deprecated](https://img.shields.io/badge/-deprecated-red)<br>Replaced by `feature_flag.provider.id`. |

Reviewer 2: ```suggestion
| <a id="feature-flag-provider-name" href="#feature-flag-provider-name">`feature_flag.provider_name`</a> | string | Deprecated, use `feature_flag.provider.id` instead. | `Flag Manager` | ![Deprecated](https://img.shields.io/badge/-deprecated-red)<br>Replaced by `feature_flag.system`. |
```

| <a id="feature-flag-context-id" href="#feature-flag-context-id">`feature_flag.context.id`</a> | string | The unique identifier for the flag evaluation context. For example, the targeting key. | `5157782b-2203-4c80-a857-dbbd5e7761db` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| <a id="feature-flag-evaluation-error-message" href="#feature-flag-evaluation-error-message">`feature_flag.evaluation.error.message`</a> | string | A message explaining the nature of an error occurring during flag evaluation. | `Flag `header-color` expected type `string` but found type `number`` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| <a id="feature-flag-evaluation-reason" href="#feature-flag-evaluation-reason">`feature_flag.evaluation.reason`</a> | string | The reason code which shows how a feature flag value was determined. | `static`; `targeting_match`; `error`; `default` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| <a id="feature-flag-key" href="#feature-flag-key">`feature_flag.key`</a> | string | The lookup key of the feature flag. | `logo-color` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

> | <a id="feature-flag-set-id" href="#feature-flag-set-id">`feature_flag.set.id`</a> | string | The identifier of the [flag set](https://openfeature.dev/specification/glossary/#flag-set) to which the feature flag belongs. | `proj-1`; `ab98sgs`; `service1/dev` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 3: Were people aligned on `feature_flag.set.id`? In our service, we use a "label" which a string that groups flags together but doesn't offer anything more. 

The `id` here makes it seem like `feature_flag.set` is a larger concept that could have other fields on it. I'm unfamiliar with what those would be. 

Did we consider something flatter? Like perhaps `| feature_flag.set | string |`

Reviewer 3: I suppose we also offer tags which some users use to group flags. An array here could offer more flexibility for grouping  `| feature_flag.tags | string[] |`. I know this isn't defined in OpenFeature but I also can't tell if OpenFeature has a limit for how many sets a flag can be in. This would better support multiple.

Author: I think it's just because of the history of renames. It was environment id when we thought we might multiple fields on environment, then we renamed environment to flag set. There wasn't really a conscious choice to keep the `id` part for flag set. I think we could definitely consider flattening to `feature_flag.set`. @beeme1mr do you think we're likely to have anything other than id in the future?

Author: It sounds to me like your `label` concept is a set id. Am I misunderstanding? I don't see any strong reason to flatten in this instance. We're using id in a lot of places and this is consistent with that, and it allows the potential for expansion in the future. One example of a future expansion might be `feature_flag.set.source` which could be `disk`, `network`, `git`, or something else. 

If your flag label is not a flag set ID can you please provide an example to help me understand?

Reviewer 3: > It sounds to me like your `label` concept is a set id.

Correct. In the systems I'm familiar with, 'set' is not an object that justifies a true 'id' or other additional fields. It's only a string that relates it to other flags.

In Azure, we offer a `string` or `string[]` to group flags together (or support prefixes). In Launch Darkly (I believe they use tags which is `string[]`). 

I'm trying to learn about how OpenFeature implements sets- but I don't see it in any of the SDKs I've inspected. I only see it mentioned in its glossary- which makes it sound like it's just a string. Do you or anyone else have a link to where sets are used or more defined in OpenFeature? I checked the .net, go, and node sdks and don't see anything with "set". It's not a field on their eval events either. 

> `feature_flag.set.source` which could be `disk`, `network`, `git`, or something else.

Thank you, this is valid justification. To me- this wouldn't belong on the evaluation event. And if it did- it would belong on the flag itself> `feature_flag.source` or possibly `feature_flag.system.source`.

Author: It _can_ map to environment, or to something else. A set of flags may be the `dev` flags, but it may also be `frontend` flags. It's really up to the user and different backends have different ideas of when to group flags together. That's why we decided to keep the grouping name fairly generic with `set`.

# Feature Flag

- [Feature Flag Attributes](#feature-flag-attributes)
- [Deprecated Feature Flag Attributes](#deprecated-feature-flag-attributes)

## Feature Flag Attributes

This document defines attributes for Feature Flags.

| Attribute | Type | Description | Examples | Stability |
|---|---|---|---|---|
| <a id="feature-flag-context-id" href="#feature-flag-context-id">`feature_flag.context.id`</a> | string | The unique identifier for the flag evaluation context. For example, the targeting key. | `5157782b-2203-4c80-a857-dbbd5e7761db` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

> | <a id="feature-flag-error-code" href="#feature-flag-error-code">`feature_flag.error.code`</a> | string | An error code which shows why a flag evaluation failed | `provider_not_ready`; `targeting_key_missing`; `provider_fatal`; `general` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 4: ```suggestion
| <a id="feature-flag-error-code" href="#feature-flag-error-code">`feature_flag.error.code`</a> | string | An error code which denotes why a flag evaluation failed | `provider_not_ready`; `targeting_key_missing`; `provider_fatal`; `general` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```

> | <a id="feature-flag-error-message" href="#feature-flag-error-message">`feature_flag.error.message`</a> | string | A message explaining why an error occurred during flag evaluation. | `Flag `header-color` expected type `string` but found type `number`` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 4: ```suggestion
| <a id="feature-flag-error-message" href="#feature-flag-error-message">`feature_flag.error.message`</a> | string | A message explaining the nature of an error occurring during flag evaluation | `Flag `header-color` expected type `string` but found type `number`` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```

A bit of a nit, but "explaining why an error occurred" seems slightly inaccurate. Messages describe errors, generally, not just "why" they occurred.

| <a id="feature-flag-key" href="#feature-flag-key">`feature_flag.key`</a> | string | The lookup key of the feature flag. | `logo-color` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| <a id="feature-flag-reason" href="#feature-flag-reason">`feature_flag.reason`</a> | string | The reason code which shows how a feature flag value was determined. | `static`; `targeting_match`; `error`; `default` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

> | <a id="feature-flag-set-id" href="#feature-flag-set-id">`feature_flag.set.id`</a> | string | The identifier of the [flag set](https://openfeature.dev/specification/glossary/#flag-set) which the feature flag belongs to in a flag management system. | `proj-1`; `ab98sgs`; `service1/dev` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 4: ```suggestion
| <a id="feature-flag-set-id" href="#feature-flag-set-id">`feature_flag.set.id`</a> | string | The identifier of the [flag set](https://openfeature.dev/specification/glossary/#flag-set) to which the feature flag belongs. | `proj-1`; `ab98sgs`; `service1/dev` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```

Nit: I think I would leave out the "management system" bit; "management system" implies a robustness that we should not assume, IMO.
